File: src/render/transitions/trim-xfade-concat.ts
import path from 'node:path';
import { RenderError } from '../../errors.js';
import { ffmpegXfadeConcat } from './xfade-concat.js';
import {
  collectClips,
  normalizeClips,
  type TransitionHandlerArgs,
  type TransitionHandlerResult,
} from './shared.js';

export function handleTrimXfadeConcat(
  args: TransitionHandlerArgs,
): TransitionHandlerResult {
  const { rootDir, ffmpegPath, ffprobePath, request, videoInfo, qualPreset, resPreset, task, outDir, outFile } = args;

  console.log(`\n[1/3] 收集并校验视频片段...`);

  const trimStart = typeof request.variables.trim_start === 'number'
    ? request.variables.trim_start
    : 2;
  const transitionDuration = typeof request.variables.transition_duration === 'number'
    ? request.variables.transition_duration
    : 0.5;

  const allClips = collectClips(request.variables, request.templateInfo, videoInfo);
  const clips = allClips.filter((clip) => {
    if (clip.duration > 0 && clip.duration <= trimStart) {
      console.warn(`  ⚠ 跳过 ${clip.key}：时长 ${clip.duration.toFixed(1)}s 不足 ${trimStart}s`);
      return false;
    }
    return true;
  });

  if (clips.length === 0) {
    throw new RenderError('没有可用的视频片段（全部片段时长均不足裁剪长度或未提供）');
  }

  console.log(`  共 ${clips.length} 个片段，裁去前 ${trimStart}s，前段淡出转场时长 ${transitionDuration}s`);

  const normalized = normalizeClips(
    rootDir,
    ffmpegPath,
    clips,
    qualPreset,
    resPreset,
  );

  const outputPath = path.join(outDir, outFile);
  ffmpegXfadeConcat(
    ffmpegPath,
    ffprobePath,
    normalized.clips,
    outputPath,
    qualPreset,
    task,
    resPreset,
    transitionDuration,
    trimStart,
  );

  return {
    cleanup: normalized.cleanup,
    outputPath,
  };
}

File: videocut/render/transitions/trim_xfade_concat.py
from __future__ import annotations

from pathlib import Path

from videocut.errors import RenderError
from videocut.log import get_logger
from videocut.render.transitions.shared import (
    TransitionHandlerArgs,
    TransitionHandlerResult,
    collect_clips,
    normalize_clips,
)
from videocut.render.transitions.xfade_concat import ffmpeg_xfade_concat


logger = get_logger(__name__)


def handle_trim_xfade_concat(args: TransitionHandlerArgs) -> TransitionHandlerResult:
    logger.info("[1/3] Collecting input clips...")
    trim_start = (
        float(args.request.variables["trim_start"])
        if isinstance(args.request.variables.get("trim_start"), (int, float))
        else 2.0
    )
    transition_duration = (
        float(args.request.variables["transition_duration"])
        if isinstance(args.request.variables.get("transition_duration"), (int, float))
        else 0.5
    )
    all_clips = collect_clips(args.request.variables, args.request.template_info, args.video_info)
    clips = []
    for clip in all_clips:
        if clip.duration > 0 and clip.duration <= trim_start:
            logger.warning("  skipping %s, duration %.1fs <= %.1fs", clip.key, clip.duration, trim_start)
            continue
        clips.append(clip)
    if not clips:
        raise RenderError("No usable video clips after trim filtering.")
    logger.info("  %d clip(s), trim first %.1fs, front fade transition %.1fs", len(clips), trim_start, transition_duration)
    normalized_clips, cleanup = normalize_clips(
        args.root_dir,
        args.ffmpeg_path,
        clips,
        args.qual_preset,
        args.res_preset,
    )
    output_path = str(Path(args.out_dir) / args.out_file)
    ffmpeg_xfade_concat(
        args.ffmpeg_path,
        normalized_clips,
        output_path,
        args.qual_preset,
        args.task,
        args.res_preset,
        transition_duration,
        trim_start=trim_start,
    )
    return TransitionHandlerResult(cleanup=cleanup, output_path=output_path)

